File: go/test/parking_record_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/parking-stgallen-sdk"
	"github.com/voxgig-sdk/parking-stgallen-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestParkingRecordEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ParkingRecord(nil)
		if ent == nil {
			t.Fatal("expected non-nil ParkingRecordEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := parking_recordBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "parking_record." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set PARKINGSTGALLEN_TEST_PARKING_RECORD_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		parkingRecordRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.parking_record", setup.data)))
		var parkingRecordRef01Data map[string]any
		if len(parkingRecordRef01DataRaw) > 0 {
			parkingRecordRef01Data = core.ToMapAny(parkingRecordRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = parkingRecordRef01Data

		// LIST
		parkingRecordRef01Ent := client.ParkingRecord(nil)
		parkingRecordRef01Match := map[string]any{}

		parkingRecordRef01ListResult, err := parkingRecordRef01Ent.List(parkingRecordRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, parkingRecordRef01ListOk := parkingRecordRef01ListResult.([]any)
		if !parkingRecordRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", parkingRecordRef01ListResult)
		}

		// LOAD
		parkingRecordRef01MatchDt0 := map[string]any{}
		parkingRecordRef01DataDt0Loaded, err := parkingRecordRef01Ent.Load(parkingRecordRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if parkingRecordRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func parking_recordBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "parking_record", "ParkingRecordTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read parking_record test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse parking_record test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"parking_record01", "parking_record02", "parking_record03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("PARKINGSTGALLEN_TEST_PARKING_RECORD_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"PARKINGSTGALLEN_TEST_PARKING_RECORD_ENTID": idmap,
		"PARKINGSTGALLEN_TEST_LIVE":      "FALSE",
		"PARKINGSTGALLEN_TEST_EXPLAIN":   "FALSE",
		"PARKINGSTGALLEN_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["PARKINGSTGALLEN_TEST_PARKING_RECORD_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["PARKINGSTGALLEN_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["PARKINGSTGALLEN_APIKEY"],
			},
			extra,
		})
		client = sdk.NewParkingStgallenSDK(core.ToMapAny(mergedOpts))
	}

	live := env["PARKINGSTGALLEN_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["PARKINGSTGALLEN_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/parking_record_entity_test.rb
# ParkingRecord entity test

require "minitest/autorun"
require "json"
require_relative "../ParkingStgallen_sdk"
require_relative "runner"

class ParkingRecordEntityTest < Minitest::Test
  def test_create_instance
    testsdk = ParkingStgallenSDK.test(nil, nil)
    ent = testsdk.ParkingRecord(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = parking_record_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "parking_record." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set PARKINGSTGALLEN_TEST_PARKING_RECORD_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    parking_record_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.parking_record")))
    parking_record_ref01_data = nil
    if parking_record_ref01_data_raw.length > 0
      parking_record_ref01_data = Helpers.to_map(parking_record_ref01_data_raw[0][1])
    end

    # LIST
    parking_record_ref01_ent = client.ParkingRecord(nil)
    parking_record_ref01_match = {}

    parking_record_ref01_list_result = parking_record_ref01_ent.list(parking_record_ref01_match, nil)
    assert parking_record_ref01_list_result.is_a?(Array)

    # LOAD
    parking_record_ref01_match_dt0 = {}
    parking_record_ref01_data_dt0_loaded = parking_record_ref01_ent.load(parking_record_ref01_match_dt0, nil)
    assert !parking_record_ref01_data_dt0_loaded.nil?

  end
end

def parking_record_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "parking_record", "ParkingRecordTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = ParkingStgallenSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["parking_record01", "parking_record02", "parking_record03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["PARKINGSTGALLEN_TEST_PARKING_RECORD_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "PARKINGSTGALLEN_TEST_PARKING_RECORD_ENTID" => idmap,
    "PARKINGSTGALLEN_TEST_LIVE" => "FALSE",
    "PARKINGSTGALLEN_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["PARKINGSTGALLEN_TEST_PARKING_RECORD_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["PARKINGSTGALLEN_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = ParkingStgallenSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["PARKINGSTGALLEN_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["PARKINGSTGALLEN_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
